### PR TITLE
bug: fixing client to handle paginated responses from Blackboard

### PR DESCRIPTION
**Pull request by alex-sheehan-edx**
[Jira](https://openedx.atlassian.net/browse/ENT-4635)

**Background:**
Through some investigations of the API documentation, we discovered that BB does indeed paginate their responses. In testing with our local instance it seems like the default size for an instance is `200` but it is unclear if this is customizable by the customer, so to be safe all sizes should be taken into account. Before this PR, if the number of assignments within a course got too high, we would not be able to verify the existence of further assignments. This would trigger our gradebook column (Blackboard's representation of our assignments) creation process, which would fail with a unique key integrity error. As such all further grade reporting would fail to be reported. 

**Code Changes in this PR**
- The Blackboard client now checks for the existence of a `next page` within the API response
- The Blackboard client will request next pages from the customer's instance until it:
    -  a) finds the assignment gradebook column
    -  b) exhausts all pages from the API
    - c) hits the page limit (250)
- Unit tests that cover the pagination traversal

**What this PR does NOT do**
The [Jira](https://openedx.atlassian.net/browse/ENT-4635) ticket specifies that we have transmitted duplicate assignments as a result of the lack of pagination handling and thus all customer instances need to be sanitize of duplicate assessment reporting (as Canvas needed recently). Through testing I have determined that this is improbable as Blackboard, unlike Canvas, has unique key restraints on the gradebook columns making it impossible to retransmit assignments with the same `externalID`. Thus any deduplication work has been left out of this PR.

**Merge checklist:**
- [ ] Any new requirements are in the right place (do **not** manually modify the `requirements/*.txt` files)
    - `base.in` if needed in production but edx-platform doesn't install it
    - `test-master.in` if edx-platform pins it, with a matching version
    - `make upgrade && make requirements` have been run to regenerate requirements
- [ ] `make static` has been run to update webpack bundling if any static content was updated
- [ ] `./manage.py makemigrations` has been run
    - Checkout the [Database Migration](https://openedx.atlassian.net/wiki/spaces/AC/pages/23003228/Everything+About+Database+Migrations) Confluence page for helpful tips on creating migrations.
    - *Note*: This **must** be run if you modified any models.
      - It may or may not make a migration depending on exactly what you modified, but it should still be run.
    - This should be run from either a venv with all the lms/edx-enterprise requirements installed or if you checked out edx-enterprise into the src directory used by lms, you can run this command through an lms shell.
        - It would be `./manage.py lms makemigrations` in the shell.
- [ ] [Version](https://github.com/edx/edx-enterprise/blob/master/enterprise/__init__.py) bumped
- [ ] [Changelog](https://github.com/edx/edx-enterprise/blob/master/CHANGELOG.rst) record added
- [ ] Translations updated (see docs/internationalization.rst but also this isn't blocking for merge atm)

**Post merge:**
- [ ] Tag pushed and a new [version](https://github.com/edx/edx-enterprise/releases) released
    - *Note*: Assets will be added automatically. You just need to provide a tag (should match your version number) and title and description.
- [ ] After versioned build finishes in [GitHub Actions](https://github.com/edx/edx-enterprise/actions), verify version has been pushed to [PyPI](https://pypi.org/project/edx-enterprise/)
    - Each step in the release build has a condition flag that checks if the rest of the steps are done and if so will deploy to PyPi.
    (so basically once your build finishes, after maybe a minute you should see the new version in PyPi automatically (on refresh))
- [ ] PR created in [edx-platform](https://github.com/edx/edx-platform) to upgrade dependencies (including edx-enterprise)
    - This **must** be done after the version is visible in PyPi as `make upgrade` in edx-platform will look for the latest version in PyPi.
    - Note: the edx-enterprise constraint in edx-platform **must** also be bumped to the latest version in PyPi.

### CODE REVIEW

File: enterprise/__init__.py
Comment (taliaedX) [inline enterprise/__init__.py:5]: womp

File: CHANGELOG.rst
Comment (taliaedX) [inline CHANGELOG.rst:20]: 3.27.3 now. 😅

File: integrated_channels/blackboard/client.py
Comment (taliaedX) [inline integrated_channels/blackboard/client.py:584]: can you ELI5 why grade_column_id not existing affects whether we check for more pages or not. I did read the description, but I may be too hazy to comprehend it if it was mentioned there.

File: integrated_channels/blackboard/client.py
Comment (taliaedX) [inline integrated_channels/blackboard/client.py:592]: is there a point in incrementing this if we set `more_pages_present = False`
To state another way, why not stick this in the `if` statement above (and I mean the one right above this Else. The other one I understand is within a For loop)

File: integrated_channels/blackboard/client.py
Comment (taliaedX) [reply integrated_channels/blackboard/client.py:584]: Is it that we go through pages of results until we get a grade_column_id and then we stop going through the pages?

**Review by taliaedX — COMMENTED** [review]
I mostly have questions for my understanding! Thus just commenting for now, but can revisit after that and when maybe my brain functions better if you still need an additional approver.

File: integrated_channels/blackboard/client.py
Comment (edx-abolger) [inline integrated_channels/blackboard/client.py:357]: nit: If you don't use 'self', make this a static or class method

File: CHANGELOG.rst
Comment (alex-sheehan-edx) [reply CHANGELOG.rst:20]: 😡

File: enterprise/__init__.py
Comment (alex-sheehan-edx) [reply enterprise/__init__.py:5]: 👍

File: integrated_channels/blackboard/client.py
Comment (alex-sheehan-edx) [reply integrated_channels/blackboard/client.py:584]: Yeah the idea is that we either go through every page or we find the ID on a particular page. Blackboard will only include the `paging` response if there exists another page, so that line is saying "if we haven't found ID yet and if there are more pages available".

File: integrated_channels/blackboard/client.py
Comment (alex-sheehan-edx) [reply integrated_channels/blackboard/client.py:592]: TBH you're right, there's no need to increment for both cases. If we have no more pages present then there's no point in incrementing the page count. It doesn't necessarily break anything to have it there, but for the sake of readability I can change.

**Review by taliaedX — APPROVED** [review]
LGTM! Understand if you want to wait until Adrienne reviews though

File: tests/test_integrated_channels/test_blackboard/test_client.py
Comment (edx-abolger) [inline tests/test_integrated_channels/test_blackboard/test_client.py:803]: Can you add some kind of test here that we call the pages the correct number of times? I don't see a test here that actually confirms that we make that call to page 2?

File: integrated_channels/blackboard/client.py
Comment (alex-sheehan-edx) [reply integrated_channels/blackboard/client.py:367]: The field isn't displayed on the immediate UI and is more of a body div type display so I thought it lent itself to a full sentence structure. That being said I also see it being completely unnecessary and differently patterned from most of the other content we display so I'll remove it.

File: tests/test_integrated_channels/test_blackboard/test_client.py
Comment (alex-sheehan-edx) [reply tests/test_integrated_channels/test_blackboard/test_client.py:803]: `responses.RequestsMock` by default asserts that all mock created within the context must be executed before leaving. If we didn't make the call to page 2 then the test would fail with `AssertionError: Not all requests have been executed [('GET', 'https://base.url/learn/api/public/v1/courses/723/gradebook/columns')]`

Comment (edx-abolger) [reply tests/test_integrated_channels/test_blackboard/test_client.py:803]: Sweet, TIL!

**Review by edx-abolger — APPROVED** [review]

**Pull request merged**
